File: day3/part1.py
import fileinput

ONE = "1"
ZERO = "0"

def get_counter(inputs):
    counter = [0] * len(inputs[0])
    for line in inputs:
        for index, bit in enumerate(line):
            counter[index] += 1 if bit == ONE else -1
    return counter

counter = get_counter([line.strip() for line in fileinput.input()])
gamma = int("".join(ONE if index >= 0 else ZERO for index in counter), 2)
epsilon = int("".join(ONE if index < 0 else ZERO for index in counter), 2)
print(gamma * epsilon)

File: day3/part1.rb
ONE = "1"
ZERO = "0"

lines = ARGF.each_line.map(&:chomp).to_a
counts = Array.new(lines.first.size) { 0 }
lines.each do |line|
  line.each_char.with_index do |c, i|
    counts[i] += 1 if c == ONE
  end
end
majority = lines.size / 2
gamma = counts.map { |n| n >= majority ? ONE : ZERO }.join.to_i(2)
epsilon = counts.map { |n| n >= majority ? ZERO : ONE }.join.to_i(2)

puts gamma * epsilon
